### PR TITLE
SC-2033 Create APIs to update and read consent

### PR DESCRIPTION
**Pull request by reshmi-nair**

### CODE REVIEW

File: actors/sunbird-lms-mw/actors/sunbird-utils/sunbird-platform-core/common-util/src/main/java/org/sunbird/common/models/util/JsonKey.java [hunk 1]
```diff
@@ -687,6 +687,20 @@
...
-
+  public static final String CONSUMERID = "consumerId";
+  public static final String OBJECTID = "objectId";
+  public static final String CONSUMERTYPE = "consumerType";
+  public static final String OBJECTTYPE = "objectType";
+  public static final String CONSENT_EXPIRY_IN_DAYS = "consent_expiry_in_days";
+  public static final String EXPIRY = "expiry";
+  public static final String CONSENT = "consent";
+  public static final String CONSENTS = "consents";
+  public static final String CONSUMER_ID = "consumer_id";
```
Comment (indrajra) [inline actors/sunbird-lms-mw/actors/sunbird-utils/sunbird-platform-core/common-util/src/main/java/org/sunbird/common/models/util/JsonKey.java:698]: an intent on the var name can add value - just a suggestion.

File: actors/sunbird-lms-mw/actors/sunbird-utils/sunbird-platform-core/common-util/src/main/resources/externalresource.properties [hunk 2]
```diff
@@ -162,3 +162,4 @@ adminutil_sign_endpoint = v1/sign/payload
 self_declared_mandatory_fields = Diksha UUID,Status,State provided ext. ID,Channel,Persona
 self_declared_optional_fields = School Name,School UDISE ID,Email ID,Phone number,Error Type
 enable_captcha=true
+consent_expiry_in_days=3
```
Comment (indrajra) [inline actors/sunbird-lms-mw/actors/sunbird-utils/sunbird-platform-core/common-util/src/main/resources/externalresource.properties:165]: Would expect this to be made available in hours instead. May be ok for now. Let @maheshkumargangula advise.

File: actors/sunbird-lms-mw/actors/user/src/main/java/org/sunbird/user/actors/UserConsentActor.java [hunk 3]
```diff
@@ -43,16 +60,22 @@ public void onReceive(Request request) throws Throwable {
...
 
     private void getUserConsent(Request request) {
+        Map<String, Object> consent = (Map<String, Object>) request.getRequest().getOrDefault(JsonKey.CONSENT, new HashMap<String, Object>());
+        Map<String, Object> filters = (Map<String, Object>) consent.getOrDefault(JsonKey.FILTERS, new HashMap<String, Object>());
+        String userId = (String) filters.getOrDefault(JsonKey.USER_ID, "");
+        String consumerId = (String) filters.getOrDefault(JsonKey.CONSUMERID, "");
+        String objectId = (String) filters.getOrDefault(JsonKey.OBJECTID, "");
+
+        Map<String, Object> getConsentFromDBReq = new HashMap<String, Object>();
+        getConsentFromDBReq.put(JsonKey.CONSENT_USER_ID, userId);
```
Comment (indrajra) [inline actors/sunbird-lms-mw/actors/user/src/main/java/org/sunbird/user/actors/UserConsentActor.java:70]: missing if clause? 
Calling with null or empty consumerId, objectId or userId must be removed. If this is already validated at the controller layer, then this must assert and not have a default.

File: actors/sunbird-lms-mw/actors/user/src/main/java/org/sunbird/user/actors/UserConsentActor.java [hunk 4]
```diff
@@ -63,19 +86,81 @@ private void getUserConsent(Request request) {
...
 
     private void updateUserConsent(Request request) {
+        Map<String, Object> consent = (Map<String, Object>) request.getRequest().getOrDefault(JsonKey.CONSENT, new HashMap<String, Object>());
+        String userId = (String) consent.getOrDefault(JsonKey.USER_ID, "");
+        if(StringUtils.isEmpty(userId)){
+            userId = (String) request.getContext().get(JsonKey.REQUESTED_BY);
+        }
+        String managedFor = (String)request.getContext().get(JsonKey.MANAGED_FOR);
+        if(StringUtils.isNotEmpty(managedFor)){
+            userId = managedFor;
```
Comment (indrajra) [inline actors/sunbird-lms-mw/actors/user/src/main/java/org/sunbird/user/actors/UserConsentActor.java:96]: looks there is a possibility for manager_for to always override user. Imagine if this header was present (un-necessarily). Better to convert to if-else, I think.

File: actors/sunbird-lms-mw/actors/user/src/main/java/org/sunbird/user/actors/UserConsentActor.java [hunk 5]
```diff
@@ -63,19 +86,81 @@ private void getUserConsent(Request request) {
...
-        String userId = (String) consent.getOrDefault("userId", "");
-        String consumerId = (String) consent.getOrDefault("consumerId", "");
-        String objectId = (String) consent.getOrDefault("objectId", "");
-        consentCache.put(getKey(userId, consumerId, objectId), consent);
-        response.put("consent", new HashMap<String, Object>() {{ put("userId", userId);}});
-        response.put("message", "User Consent updated successfully.");
+        Map consentResponse = new HashMap<String, Object>();
+        consentResponse.put(JsonKey.USER_ID, userId);
+        response.put(JsonKey.CONSENT, consentResponse);
+        response.put(JsonKey.MESSAGE, "User Consent updated successfully.");
```
Comment (indrajra) [inline actors/sunbird-lms-mw/actors/user/src/main/java/org/sunbird/user/actors/UserConsentActor.java:109]: Stringify message in a constant.

File: actors/sunbird-lms-mw/actors/user/src/main/java/org/sunbird/user/actors/UserConsentActor.java [hunk 6]
```diff
@@ -63,19 +86,81 @@ private void getUserConsent(Request request) {
...
+                        userId, TelemetryEnvKey.USER, JsonKey.UPDATE, null);
+        TelemetryUtil.telemetryProcessingCall(
+                consent, targetObject, correlatedObject, request.getContext());
     }
 
     private String getKey (String userId, String consumerId, String objectId) {
         return "usr-consent:" + userId + ":" + consumerId + ":" + objectId;
     }
 
+    private Map<String, Object> createConsentRequest(Map<String, Object> consent, RequestContext context){
```
Comment (indrajra) [inline actors/sunbird-lms-mw/actors/user/src/main/java/org/sunbird/user/actors/UserConsentActor.java:126]: this is possibly constructing a consent map, the term create is a little off the mark here.

File: actors/sunbird-lms-mw/actors/user/src/main/java/org/sunbird/user/dao/impl/UserConsentDaoImpl.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+import java.util.Map;
+
+public class UserConsentDaoImpl implements UserConsentDao {
+    private static final String TABLE_NAME = "user_consent";
+    private CassandraOperation cassandraOperation = ServiceFactory.getInstance();
+    private ObjectMapper mapper = new ObjectMapper();
+
+    private static UserConsentDao consentDao = null;
+
+    public static UserConsentDao getInstance() {
```
Comment (indrajra) [inline actors/sunbird-lms-mw/actors/user/src/main/java/org/sunbird/user/dao/impl/UserConsentDaoImpl.java:23]: should this be really singleton? I dont think so.

File: actors/sunbird-lms-mw/actors/user/src/main/java/org/sunbird/user/actors/UserConsentActor.java (hunk 4 above)
Comment (reshmi-nair) [reply actors/sunbird-lms-mw/actors/user/src/main/java/org/sunbird/user/actors/UserConsentActor.java:96]: userId is not mandatory in request. So if managedFor is present that should get preference rite?

File: actors/sunbird-lms-mw/actors/user/src/main/java/org/sunbird/user/actors/UserConsentActor.java (hunk 5 above)
Comment (reshmi-nair) [reply actors/sunbird-lms-mw/actors/user/src/main/java/org/sunbird/user/actors/UserConsentActor.java:109]: Done!

File: actors/sunbird-lms-mw/actors/user/src/main/java/org/sunbird/user/actors/UserConsentActor.java (hunk 6 above)
Comment (reshmi-nair) [reply actors/sunbird-lms-mw/actors/user/src/main/java/org/sunbird/user/actors/UserConsentActor.java:126]: Done!

File: actors/sunbird-lms-mw/actors/user/src/main/java/org/sunbird/user/actors/UserConsentActor.java (hunk 3 above)
Comment (reshmi-nair) [reply actors/sunbird-lms-mw/actors/user/src/main/java/org/sunbird/user/actors/UserConsentActor.java:70]: Validated at controller layer. So removing defaults.

File: actors/sunbird-lms-mw/actors/sunbird-utils/sunbird-platform-core/common-util/src/main/java/org/sunbird/common/models/util/JsonKey.java (hunk 1 above)
Comment (reshmi-nair) [reply actors/sunbird-lms-mw/actors/sunbird-utils/sunbird-platform-core/common-util/src/main/java/org/sunbird/common/models/util/JsonKey.java:698]: Done

File: actors/sunbird-lms-mw/actors/sunbird-utils/sunbird-platform-core/common-util/src/main/resources/externalresource.properties (hunk 2 above)
Comment (reshmi-nair) [reply actors/sunbird-lms-mw/actors/sunbird-utils/sunbird-platform-core/common-util/src/main/resources/externalresource.properties]: updated to 100 days as per mahesh's suggestion

**Review by maheshkumargangula — APPROVED** [review]

**Pull request merged**